src/data/int/log.lean

```diff
@@ -0,0 +1,238 @@
+begin
+  have h1b' : 1 ≤ (b : R) := by exact_mod_cast hb.le,
+  rw log,
+  split_ifs,
+  { obtain ⟨a, rfl | rfl⟩ := x.eq_coe_or_neg,
+    { simp only [zpow_coe_nat, ←nat.cast_pow, ← nat.le_floor_iff hy.le, int.coe_nat_le],
+      exact nat.pow_le_iff_le_log hb (nat.floor_pos.mpr h) },
+    { have hna : -(a : ℤ) ≤ 0 := neg_nonpos.mpr (int.coe_nat_nonneg _),
+      refine iff_of_true _ _,
+      { exact (zpow_le_one_of_nonpos h1b' hna).trans h },
+      { exact hna.trans (int.coe_nat_nonneg _) } } },
+  { obtain ⟨a, rfl | rfl⟩ := x.eq_coe_or_neg,
+    { refine iff_of_false (mt (le_trans _) h) (λ ha, _),
+      { exact one_le_zpow_of_nonneg h1b' (int.coe_nat_nonneg _)},
+      { have := (int.coe_nat_nonneg _).trans ha,
+        rw [neg_nonneg, coe_nat_nonpos_iff] at this,
+        refine (nat.clog_pos hb _).ne' this,
+        have := (one_lt_inv hy (not_le.mp h)).trans_le (nat.le_ceil _),
+        rw nat.one_lt_cast at this,
+        exact this } },
+    { rw [zpow_neg₀, zpow_coe_nat, neg_le_neg_iff, int.coe_nat_le,
+        inv_le (pow_pos (zero_lt_one.trans_le h1b') _) hy, ←nat.le_pow_iff_clog_le hb,
+        ←nat.cast_pow, nat.ceil_le] }, },
+end
+
+/-- The least power of `b` such that `r ≤ b ^ log b r`. -/
+def clog (b : ℕ) (r : R) : ℤ :=
+if 1 ≤ r then
+  nat.clog b (nat.ceil r)
+else
+  -nat.log b (nat.floor r⁻¹)
+
+@[simp] lemma clog_nat_cast (b : ℕ) (n : ℕ) : clog b (n : R) = nat.clog b n :=
+begin
+  rw clog,
+  cases n,
+  { simp [nat.clog_zero_right] },
+  { simp [←nat.cast_succ] }
+end
+
+lemma clog_of_left_le_one {b : ℕ} (hb : b ≤ 1) (n : R) : clog b n = 0 :=
+begin
+  rw clog,
+  split_ifs with hn,
+  { rw [nat.clog_of_left_le_one hb, int.coe_nat_zero] },
+  { rw [nat.log_of_left_le_one hb, int.coe_nat_zero, neg_zero] },
+end
+
+lemma clog_of_right_le_zero {n : R} (hn : n ≤ 0) (b : ℕ) : clog b n = 0 :=
+begin
+  rw [clog, if_neg (hn.trans_lt zero_lt_one).not_le, neg_eq_zero, int.coe_nat_eq_zero,
+    nat.log_eq_zero_iff],
+  cases le_or_lt b 1 with hb hb,
+  { exact or.inr hb },
+  { refine or.inl (lt_of_le_of_lt _ hb),
+    exact nat.floor_le_one_of_le_one ((inv_nonpos.2 hn).trans zero_le_one) },
+end
+
+lemma clog_inv (b : ℕ) (r : R) : clog b r⁻¹ = -log b r :=
+begin
+  cases lt_or_le 0 r with hrp hrp,
+  { obtain hr | rfl | hr := lt_trichotomy r 1,
+    { have : 1 < r⁻¹ := one_lt_inv hrp hr,
+      rw [clog, if_pos this.le, log, if_neg hr.not_le, neg_neg] },
+    { rw [inv_one, clog, log, if_pos le_rfl, if_pos le_rfl, nat.ceil_one, nat.floor_one,
+        nat.clog_one_right, nat.log_one_right, int.coe_nat_zero, neg_zero], },
+    { have : r⁻¹ < 1 := inv_lt_one hr,
+      rw [clog, if_neg this.not_le, log, if_pos hr.le, inv_inv] } },
+  { rw [clog_of_right_le_zero (inv_nonpos.mpr hrp), log_of_right_le_zero hrp, neg_zero], },
+end
+
+lemma log_inv (b : ℕ) (r : R) : log b r⁻¹ = -clog b r :=
+by rw [←inv_inv r, clog_inv, neg_neg, inv_inv]
+
+lemma clog_eq_neg_log_inv (b : ℕ) (r : R) : clog b r = -log b r⁻¹ :=
+by rw [log_inv, neg_neg]
```

Comment:
I would say we don't need this, but at least put the simpler side on the right. I get that you will often want to rewrite the other way but I don't think that's an excuse,
```suggestion
lemma neg_log_inv (b : ℕ) (r : R) : -log b r⁻¹ = clog b r :=
by rw [log_inv, neg_neg]
```

Reply:
Happy with `neg_log_inv_eq_clog` for anyone looking for `eq_clog`? Really the point in this lemma is to turn `clog`problems into `log` problems so that we don't have to prove things twice.

Reply:
Yeah, fine by me.